lib/src/main/kotlin/com/swmansion/starknet/data/serializers/MerkleNodeSerializer.kt

```diff
@@ -0,0 +1,42 @@
+package com.swmansion.starknet.data.serializers
+
+import com.swmansion.starknet.data.types.NodeHashToNodeMappingItem
+import kotlinx.serialization.KSerializer
+import kotlinx.serialization.encoding.Decoder
+import kotlinx.serialization.encoding.Encoder
+import kotlinx.serialization.json.JsonDecoder
+import kotlinx.serialization.json.JsonEncoder
+import kotlinx.serialization.json.jsonObject
+
+internal object MerkleNodeSerializer : KSerializer<NodeHashToNodeMappingItem.MerkleNode> {
+    override val descriptor = NodeHashToNodeMappingItem.MerkleNode.serializer().descriptor
+
+    override fun deserialize(decoder: Decoder): NodeHashToNodeMappingItem.MerkleNode {
+        require(decoder is JsonDecoder)
+        val element = decoder.decodeJsonElement()
+        val jsonElement = element.jsonObject
+
+        val binaryNodeKeys = listOf("left", "right")
+        val edgeNodeKeys = listOf("path", "length", "child")
+
+        return when {
+            binaryNodeKeys.all { jsonElement.containsKey(it) } -> {
+                decoder.json.decodeFromJsonElement(NodeHashToNodeMappingItem.BinaryNode.serializer(), element)
+            }
+            edgeNodeKeys.all { jsonElement.containsKey(it) } -> {
+                decoder.json.decodeFromJsonElement(NodeHashToNodeMappingItem.EdgeNode.serializer(), element)
+            }
+            else -> throw IllegalArgumentException("Invalid MerkleNode JSON object: $jsonElement")
+        }
+    }
```

Comment:
note: We could potentially give an error with more meaningful message (saying which fields are missing and if it's binary/edge node). However, in case of json with keys that belong both to binary and edge node
e.g.
```
{
    "left": ...,
    "path": ...
}
```
we can't explicitly decide which type it is, i.e. we can't give type-specific error. I'm not fully sure if such logic is worth introducing.